### ISSUE TITLE
use only n_models to initialize model sets

### ISSUE THREAD

**Issue by nden** [astropy/astropy#7223]
Currently model sets are initialized either by passing `n_models=n` or if `n_models` is not passed by inferring the number of models based on coefficients' shape and `model_set_axis`.
This makes code more complex. I think it's not a big burden on the user to pass `n_models` to the initializer when they want a model set so I propose to initialize a model set only when `n_models` is passed.

@jehturner @perrygreenfield (since you've both looked at mdoel sets).

**Comment by vermashresth** [comment]
Hi @nden 
I would like to work on this. Would this require some changes in the `_initialize_parameters` function in [here](https://github.com/astropy/astropy/blob/master/astropy/modeling/core.py#L1709:L1840).
Furthermore, after this is implemented, should something like
```python
g = models.Gaussian1D(amplitude=[1, 2], mean=[0, 0], stddev=[0.1, 0.2])
``` 
result in error or a valid model set of size 1?

**Comment by nden** [comment]
Hi @vermashresth, The example you give above actually creates one model with array parameters. There are issues with this type of initialization but I don't want to deal with this yet (I think it's not clear at the moment what needs to be done).

There are two other ways to initialize a model set:

- passing `n_models` and optionally `model_set_axis`:

`g = models.Gaussian1D(amplitude=[[1, 2]], mean=[[0, 0]], stddev=[[0.1, 0.2]], n_models=2, model_set_axis=1)`

- passing only `model_set_axis` in which case the number of models is inferred from the shape of the first parameter

`g = models.Gaussian1D(amplitude=[[1, 2]], mean=[[0, 0]], stddev=[[0.1, 0.2]], model_set_axis=1)`

This issues is about eliminating the second option. I believe the modifications still need to be in the `_initialize_parameters` function. The goal is to simplify the code. 

If you are still interested give it a try. I haven't looked at the details of this yet and am sorry if I don't respond fast enough.

**Comment by vermashresth** [comment]
Hey @nden 
I have proposed an inital PR implementing the changes but it doesn't seem to reduce the complexity of the code much. We can't remove the iterating over parmeters part because that is still needed to check if parameter shape is consistent with the given value of `n_models`. So, figuring out `n_models` from parametrs is just one extra step after this. Moreover, changing this would require updating a lot of tests in modelling. Is it fine or perhaps im missinng something?

**Comment by hamogu** [comment]
@nden: I agree with your suggestion (make `n_models` required for model sets). And I would also ask to broadcast default values to the right dimension when `n_models` is given.
E.g. the following fails:
```
import numpy as np
from astropy.modeling import models
n = 10
g_init = models.Gaussian2D(n_models=n)

---------------------------------------------------------------------------
InputParameterError                       Traceback (most recent call last)
<ipython-input-38-4e19f2952a31> in <module>
      1 from astropy.modeling import models
      2 n = 10
----> 3 g_init = models.Gaussian2D(n_models=n)

/nfs/melkor/d1/guenther/soft/anaconda/envs/py3/lib/python3.7/site-packages/astropy/modeling/functional_models.py in __init__(self, amplitude, x_mean, y_mean, x_stddev, y_stddev, theta, cov_matrix, **kwargs)
    310         super().__init__(
    311             amplitude=amplitude, x_mean=x_mean, y_mean=y_mean,
--> 312             x_stddev=x_stddev, y_stddev=y_stddev, theta=theta, **kwargs)
    313 
    314     @property

/nfs/melkor/d1/guenther/soft/anaconda/envs/py3/lib/python3.7/site-packages/astropy/modeling/core.py in __init__(self, meta, name, *args, **kwargs)
    682         # Parameter values must be passed in as keyword arguments in order to
    683         # distinguish them
--> 684         self._initialize_parameters(args, kwargs)
    685         self._initialize_unit_support()
    686 

/nfs/melkor/d1/guenther/soft/anaconda/envs/py3/lib/python3.7/site-packages/astropy/modeling/core.py in _initialize_parameters(self, args, kwargs)
   1816                         "at least {0} for model_set_axis={1} (the value "
   1817                         "given for {2!r} is only {3}-dimensional)".format(
-> 1818                             min_ndim, model_set_axis, name, param_ndim))
   1819 
   1820                 max_ndim = max(max_ndim, param_ndim)

InputParameterError: All parameter values must be arrays of dimension at least 1 for model_set_axis=0 (the value given for 'amplitude' is only 0-dimensional)
```

Instead, I have to repeat the definition of the input parameters and make every single one to be the right size, even those that have defaults:
```
g_init = models.Gaussian2D(amplitude=np.ones(n), x_mean=np.ones(n), y_mean=np.ones(n),
                           x_stddev=[None] * n, y_stddev=[None] * n, theta=[None] * n,
                           n_models=n)
```
That's a lot of extra work just to get a set of models with default parameters.
(The goal here is of course to fit the models, thus it's OK if they all start with the same values. After the fit, they will all be different. (And before you ask: I discovered later that there currently is not way to fit this model set, because only linear fitters support fitting model sets and this model is not linear. Nevertheless, the general problem is the same for linear fitters plus I hope that we will eventually implement non linear fitters that can deal with sets.))

**Comment by saimn** [comment]
@hamogu - See #6096 for the broadcast question, and #3670 for non-linear fitters.

**Comment by nden** [comment]
@hamogu @saimn Thanks for the feedback. This is the intended direction of development although it has stalled for now waiting on #7945 to be resolved. If you have time/interest to review #7945 please do so. I believe code clean up of that PR is coming up in the next weeks though  implementation is essentially finished.
